draft-ietf-tls-external-psk-importer.md

```diff
@@ -79,156 +80,151 @@ when, and only when, they appear in all capitals, as shown here.
+an ImportedIdentity structure as follows:
 
 ~~~
-   struct {
-       opaque external_identity<1...2^16-1>;
-       opaque label<0..2^8-1>;
-       opaque context<0..2^16-1>;
-       HashAlgorithm hash;
-   } ImportedIdentity;
+uint8 KDFID[2];
+
+struct {
+   opaque external_identity<1...2^16-1>;
+   opaque context<0..2^16-1>;
+   uint16 target_protocol;
+   KDFID target_kdf;
+} ImportedIdentity;
 ~~~
 
-[[TODO: An alternative design might combine label and hash into the same field so that future
-protocols which don't have a notion of HashAlgorithm don't need this field.]]
+The list of KDFID values is maintained by IANA as described in {{IANA}}. External PSKs MUST NOT
+be imported for versions of (D)TLS 1.2 or prior versions. See {{rollout}} for discussion on
+how imported PSKs for TLS 1.3 and non-imported PSKs for earlier versions co-exist for incremental
+deployment.
 
-ImportedIdentity.label MUST be bound to the protocol for which the key is imported. Thus,
-TLS 1.3 and QUICv1 {{!I-D.ietf-quic-transport}} MUST use "tls13" as the label. Similarly, TLS 1.2 and
-all prior TLS versions should use "tls12" as ImportedIdentity.label, as well as SHA256 as ImportedIdentity.hash.
-Note that this means future versions of TLS will increase the number of PSKs derived from an external PSK.
+ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.
+If the EPSK is a key derived from some other protocol or sequence of protocols,
+ImportedIdentity.context MUST include a channel binding for the deriving protocols
+{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included
+in ImportedIdentity.context {{CCB}}.
 
-ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
-from some other protocol or sequence of protocols, ImportedIdentity.context MUST include a channel binding for the deriving protocols
-{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included in ImportedIdentity.context [CCB].
+ImportedIdentity.target_protocol MUST be the (D)TLS protocol version for which the
+PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!I-D.ietf-quic-transport}}
+MUST use 0x0304. Note that this means future versions of TLS will increase the number of PSKs
+derived from an external PSK.
 
-A unique and imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:
+An imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:
 
 ~~~
    epskx = HKDF-Extract(0, epsk)
    ipskx = HKDF-Expand-Label(epskx, "derived psk",
-                             Hash(ImportedIdentity), Hash.length)
+                             Hash(ImportedIdentity), L)
 ~~~
 
-[[TODO: The length of ipskx MUST match that of the corresponding and supported ciphersuites.]]
+L is the length of the hash function associated with ImportedIdentity.target_kdf, as this
```

Comment:
This suddenly assumes that the target KDF has a hash function.  You might say that this value is defined by the target KDF, and for those that are hash-based it is the length of the hash function output.

Reply:
Fixed!